numpy/core/numeric.py

```diff
@@ -2442,6 +2445,16 @@ def array_equal(a1, a2, equal_nan=False):
         return False
     if not equal_nan:
         return bool(asarray(a1 == a2).all())
+    cannot_have_nan = (_dtype_cannot_hold_nan(a1.dtype) and
+                       _dtype_cannot_hold_nan(a2.dtype))
+    if cannot_have_nan:
+        if a1 is a2:
+            return True
+        return bool(asarray(a1 == a2).all())
```

Comment:
```suggestion
        return bool((a1 == a2).all())
```
Since `a1` and `a2` have already been cast with `asarray`

Reply:
I'm not confident in removing the asarray, as I'm not sure why it's here ([Chesterton's fence](https://en.wikipedia.org/wiki/G._K._Chesterton#Chesterton's_fence)):
 - it is present on line 2477 – before this patch, 
 - and on line 2461 – last return of this function also present before this patch. 


From what I can tell this was introduced in #3322 (2013), long after this function (2006) see [this comment](https://github.com/numpy/numpy/pull/3322#discussion_r4183721).

>  If the two operands are 0d arrays, the == operator will return a bool object instead of an array, in which case the call to all() will fail.

Reply:
I think this is still ok as now, 0d arrays comparison return np.True_ / np.False, but then if I remove I might as well remove also in the existing branches right ?

Reply:
It may be required for object array support, although ideall there is a test for that that fails if you remove it.

Reply:
Ok, I pushed a commit to try.